File: old/src/Stealth/WindowCloaking.m
//
//  WindowCloaking.m
//  StealthKit
//
//  Created on Phase 4: Stealth Features Implementation
//

#import "WindowCloaking.h"

@implementation WindowCloaking

+ (void)applyCloakingToWindow:(NSWindow *)window {
    if (!window) {
        NSLog(@"StealthKit: Cannot apply cloaking to nil window");
        return;
    }
    
    NSLog(@"StealthKit: Applying cloaking to window: %@", window);
    
    // Primary screen capture evasion
    window.sharingType = NSWindowSharingNone;
    
    // Configure window level for stealth operation
    [self applyStealthWindowLevel:window];
    
    // Configure collection behavior
    [self configureStealthCollectionBehavior:window];
    
    // Additional privacy configurations
    window.hidesOnDeactivate = NO;
    window.canHide = YES;
    
    // Prevent window from being included in window lists
    window.excludedFromWindowsMenu = YES;
    
    NSLog(@"StealthKit: Cloaking applied successfully");
}

+ (void)configureWebViewForStealth:(WKWebView *)webView {
    if (!webView) {
        NSLog(@"StealthKit: Cannot configure nil web view for stealth");
        return;
    }
    
    NSLog(@"StealthKit: Configuring web view for stealth operation");
    
    WKWebViewConfiguration *config = webView.configuration;
    
    // Use non-persistent data store for maximum privacy
    config.websiteDataStore = [self createStealthDataStore];
    
    // Disable media capture
    config.allowsAirPlayForMediaPlayback = NO;
    
    // Configure preferences for privacy
    WKPreferences *preferences = config.preferences;
    preferences.javaScriptCanOpenWindowsAutomatically = NO;
    
    // Additional privacy settings
    if (@available(macOS 11.0, *)) {
        config.limitsNavigationsToAppBoundDomains = NO;
    }
    
    NSLog(@"StealthKit: Web view stealth configuration complete");
}

+ (void)removeCloakingFromWindow:(NSWindow *)window {
    if (!window) {
        return;
    }
    
    NSLog(@"StealthKit: Removing cloaking from window: %@", window);
    
    // Restore normal sharing behavior
    window.sharingType = NSWindowSharingReadOnly;
    
    // Restore normal window level
    window.level = NSNormalWindowLevel;
    
    // Restore normal collection behavior
    window.collectionBehavior = NSWindowCollectionBehaviorDefault;
    
    // Restore window menu inclusion
    window.excludedFromWindowsMenu = NO;
    
    NSLog(@"StealthKit: Cloaking removed successfully");
}

+ (BOOL)windowHasCloaking:(NSWindow *)window {
    if (!window) {
        return NO;
    }
    
    return window.sharingType == NSWindowSharingNone;
}

+ (void)applyStealthWindowLevel:(NSWindow *)window {
    if (!window) {
        return;
    }
    
    window.level = NSFloatingWindowLevel;
    
    NSLog(@"StealthKit: Stealth window level applied");
}

+ (void)configureStealthCollectionBehavior:(NSWindow *)window {
    if (!window) {
        return;
    }
    
    // Configure collection behavior for stealth operation
    NSWindowCollectionBehavior behavior = NSWindowCollectionBehaviorDefault;
    
    // Allow window to join all spaces for better stealth operation
    behavior |= NSWindowCollectionBehaviorCanJoinAllSpaces;
    
    // Manage window in Exposé
    behavior |= NSWindowCollectionBehaviorParticipatesInCycle;
    
    // Allow full screen if needed
    if (@available(macOS 10.7, *)) {
        behavior |= NSWindowCollectionBehaviorFullScreenAuxiliary;
    }
    
    window.collectionBehavior = behavior;
    
    NSLog(@"StealthKit: Stealth collection behavior configured");
}

+ (WKWebsiteDataStore *)createStealthDataStore {
    // Create a non-persistent data store for maximum privacy
    WKWebsiteDataStore *dataStore = [WKWebsiteDataStore nonPersistentDataStore];
    
    NSLog(@"StealthKit: Created non-persistent website data store for stealth operation");
    
    return dataStore;
}

#pragma mark - Advanced Stealth Features

+ (void)configureAdvancedStealth:(NSWindow *)window {
    if (!window) {
        return;
    }
    
    // Additional advanced stealth configurations can be added here
    
    // Prevent window from appearing in screenshots (primary method)
    window.sharingType = NSWindowSharingNone;
    
    // Configure window to be less detectable
    window.displaysWhenScreenProfileChanges = NO;
    
    NSLog(@"StealthKit: Advanced stealth features configured");
}

+ (void)logStealthStatus:(NSWindow *)window {
    if (!window) {
        NSLog(@"StealthKit: Cannot log status for nil window");
        return;
    }
    
    NSLog(@"StealthKit: Window Stealth Status:");
    NSLog(@"  - Sharing Type: %ld", (long)window.sharingType);
    NSLog(@"  - Window Level: %ld", (long)window.level);
    NSLog(@"  - Collection Behavior: %lu", (unsigned long)window.collectionBehavior);
    NSLog(@"  - Excluded from Windows Menu: %@", window.excludedFromWindowsMenu ? @"YES" : @"NO");
    NSLog(@"  - Has Cloaking: %@", [self windowHasCloaking:window] ? @"YES" : @"NO");
}

@end

File: Sources/SwiftBrowser/Core/Services/WindowCloaking.swift
import Foundation
import AppKit
import WebKit

class WindowCloaking {
    
    // MARK: - Window Transparency
    
    static func setWindowTransparency(_ window: NSWindow, alpha: Double) {
        let clampedAlpha = max(0.1, min(1.0, alpha))
        window.alphaValue = clampedAlpha
        
        // If making window more transparent, also add background material
        if clampedAlpha < 1.0 {
            window.backgroundColor = NSColor.clear
            window.isOpaque = false
        } else {
            window.isOpaque = true
        }
    }
    
    static func getWindowTransparency(_ window: NSWindow) -> Double {
        return window.alphaValue
    }
    
    static func applyCloakingToWindow(_ window: NSWindow, isPinnedToCurrentDesktop: Bool = true) {
        configureStealthCollectionBehavior(window, isPinnedToCurrentDesktop: isPinnedToCurrentDesktop)
        applyStealthWindowLevel(window)
        configureAdvancedStealth(window)
    }
    
    static func removeCloakingFromWindow(_ window: NSWindow) {
        // Restore normal window behavior
        window.collectionBehavior = [.managed, .participatesInCycle]
        window.level = .normal
        window.sharingType = .readWrite
        
        // Remove any custom properties that were set for stealth
        window.ignoresMouseEvents = false
    }
    
    static func windowHasCloaking(_ window: NSWindow) -> Bool {
        // Check if window has stealth collection behavior applied
        return window.collectionBehavior.contains(.stationary) ||
               window.collectionBehavior.contains(.ignoresCycle) ||
               window.level != .normal
    }
    
    // MARK: - Window Level Configuration
    
    static func applyStealthWindowLevel(_ window: NSWindow) {
        // Set window level for stealth operation
        // Using normal level but with special collection behavior
        window.level = .normal
        
        // Ensure window appears correctly but with stealth properties
        window.hidesOnDeactivate = false
        window.canHide = true
    }
    
    // MARK: - Collection Behavior Configuration
    
    static func configureStealthCollectionBehavior(_ window: NSWindow, isPinnedToCurrentDesktop: Bool = true) {
        // Configure window collection behavior for stealth operation
        var behavior: NSWindow.CollectionBehavior = []
        
        // Make window invisible to screen capture but preserve Mission Control behavior
        // Check if pinned to current desktop setting is disabled
        if !isPinnedToCurrentDesktop {
            behavior.insert(.canJoinAllSpaces)
        }
        
        // Exclude from screenshots and screen recordings
        if #available(macOS 11.0, *) {
            behavior.insert(.auxiliary)
        }
        
        // Don't use .stationary or .ignoresCycle as they cause Mission Control issues
        // Instead rely on sharing type and other properties for screen capture protection
        
        window.collectionBehavior = behavior
        
        // Additional stealth configurations
        window.sharingType = .none
        window.displaysWhenScreenProfileChanges = false
    }
    
    // MARK: - Advanced Stealth Configuration
    
    static func configureAdvancedStealth(_ window: NSWindow) {
        // Apply advanced stealth features
        
        // Disable window animations that might be captured
        window.animationBehavior = .none
        
        // Configure window shadow (shadows can be captured)
        window.hasShadow = false
        
        // Additional low-level window configuration could go here
        // For now, we rely on collection behavior for window server exclusion
        
        // Set up window delegate for additional stealth handling
        if window.delegate == nil {
            window.delegate = StealthWindowDelegate.shared
        }
    }
    
    // MARK: - WebView Configuration
    
    static func configureWebViewForStealth(_ webView: WKWebView) {
        // Configure web view for maximum privacy
        let configuration = webView.configuration
        
        // Use stealth data store
        configuration.websiteDataStore = createStealthDataStore()
        
        // Note: Camera and microphone capture states are read-only
        // Privacy is handled through the non-persistent data store
        
        // Additional privacy configurations
        configuration.preferences.javaScriptCanOpenWindowsAutomatically = false
        configuration.preferences.isElementFullscreenEnabled = false
        
        // Disable media capture
        configuration.mediaTypesRequiringUserActionForPlayback = [.all]
    }
    
    static func createStealthDataStore() -> WKWebsiteDataStore {
        // Create a non-persistent website data store for maximum privacy
        let dataStore = WKWebsiteDataStore.nonPersistent()
        
        // Configure for enhanced privacy
        return dataStore
    }
    
    // MARK: - Accessory Window Configuration
    
    static func applyAccessoryWindow(_ window: NSWindow) {
        var behavior = window.collectionBehavior
        behavior.insert(.auxiliary)
        window.collectionBehavior = behavior
    }
    
    static func removeAccessoryWindow(_ window: NSWindow) {
        var behavior = window.collectionBehavior
        behavior.remove(.auxiliary)
        window.collectionBehavior = behavior
    }
    
}

// MARK: - Stealth Window Delegate

class StealthWindowDelegate: NSObject, NSWindowDelegate {
    static let shared = StealthWindowDelegate()
    
    private override init() {
        super.init()
    }
    
    func windowDidBecomeKey(_ notification: Notification) {
        // Ensure stealth properties are maintained when window becomes key
        if let window = notification.object as? NSWindow {
            let isPinnedToCurrentDesktop = StealthManager.shared.isPinnedToCurrentDesktop
            WindowCloaking.configureStealthCollectionBehavior(window, isPinnedToCurrentDesktop: isPinnedToCurrentDesktop)
        }
    }
    
    func windowDidResignKey(_ notification: Notification) {
        // Maintain stealth properties when window loses key status
        if let window = notification.object as? NSWindow {
            let isPinnedToCurrentDesktop = StealthManager.shared.isPinnedToCurrentDesktop
            WindowCloaking.configureStealthCollectionBehavior(window, isPinnedToCurrentDesktop: isPinnedToCurrentDesktop)
        }
    }
    
    func windowWillMiniaturize(_ notification: Notification) {
        // Handle window miniaturization in stealth mode
        if let window = notification.object as? NSWindow {
            // Ensure stealth properties are maintained during miniaturization
            window.collectionBehavior.insert(.stationary)
        }
    }
}
